zipper/protocols/querydispatch/querydispatch_helper.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/go-graphite/carbonapi/zipper/metadata"
+	"github.com/go-graphite/carbonapi/zipper/types"
+	protov3 "github.com/go-graphite/protocol/carbonapi_v3_pb"
+	"go.uber.org/zap"
+)
+
+func (q *QueryDispatchGroup) CreatePerFetchQueryStoreBackends(request *protov3.MultiFetchRequest) (map[types.BackendServer][]*protov3.FetchRequest, error) {
+	// Match each regex rule to each query and put backend in allow or deny groups.
+	AllowGroupPerMetric := make(map[*protov3.FetchRequest][]string)
+	DenyGroupPerMetric := make(map[*protov3.FetchRequest][]string)
+	for _, fetchReq := range request.GetMetrics() {
+		metricReq := fetchReq.GetName()
+		for _, rule := range q.backendsV2.GroupRegexRules {
+			if rule.CompiledRegex.MatchString(metricReq) {
+				if rule.Effect == "allow" {
+					AllowGroupPerMetric[&fetchReq] = types.AppendUnique(AllowGroupPerMetric[&fetchReq], rule.GroupBackendNames)
+				} else if rule.Effect == "deny" {
+					DenyGroupPerMetric[&fetchReq] = types.AppendUnique(DenyGroupPerMetric[&fetchReq], rule.GroupBackendNames)
+				} else {
+					q.logger.Warn("Unknow Effect")
+				}
+			}
+		}
+	}
+	// If a backend is denied somewhere it has to be removed from the allow list
+	for query, allowedGroupList := range AllowGroupPerMetric {
+		denyGroupList, ok := DenyGroupPerMetric[query]
+		if !ok {
+			continue
+		}
+		for index, allowedGroupName := range allowedGroupList {
+			if contains(denyGroupList, allowedGroupName) {
+
+				remove(AllowGroupPerMetric[query], index)
+				continue
+			}
+		}
+	}
+	// Reverse Order
+	queriesPerGroup := make(map[string][]*protov3.FetchRequest)
+	for query, allowedGroupList := range AllowGroupPerMetric {
+		for _, nameGroup := range allowedGroupList {
+			queriesPerGroup[nameGroup] = append(queriesPerGroup[nameGroup], query)
+		}
+	}
+
+	// Convert string representing backends to types.BackendServer & Create broadcast group for each query
+	queryToGroup := make(map[*protov3.FetchRequest]types.BackendServer)
+	for query, groupNames := range AllowGroupPerMetric {
+		var tmp []types.BackendServer
+		for _, backend := range q.backendsV2.Backends {
+			metadata.Metadata.RLock()
+			backendInit, ok := metadata.Metadata.ProtocolInits[backend.Protocol]
+			metadata.Metadata.RUnlock()
+			if !ok {
+				var protocols []string
+				metadata.Metadata.RLock()
+				for p := range metadata.Metadata.SupportedProtocols {
+					protocols = append(protocols, p)
+				}
+				metadata.Metadata.RUnlock()
+				q.logger.Error("unknown backend protocol",
+					zap.Any("backend", backend),
+					zap.String("requested_protocol", backend.Protocol),
+					zap.Strings("supported_backends", protocols),
+				)
+				continue
+			}
+			for _, groupName := range groupNames {
+				if groupName == backend.GroupName {
+					config := backend
+					for _, server := range backend.Servers {
+						config.Servers = []string{server}
+						config.GroupName = server
+						client, e := backendInit(q.logger, config, q.tldCacheDisabled)
```

Comment:
I don't get why we need to create new client for each request. Can't we create clients at init and pick them on matches? TBD